### ISSUE TITLE
[PATCH] Coverity: (unintended) sign extension avoidance

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#13776]
Migrated from [rt.perl.org#121746](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=121746) (status was 'resolved')

Searchable as RT121746$

**Comment by p5pRT** [comment]
### From @jhi

Attached\.

**Comment by p5pRT** [comment]
### From @jhi

<details><summary>0001-Fix-for-Coverity-perl5-CIDs-29069-29070-29071.patch</summary>

```diff
From d03da935fb050db149a4e83c30a1fbc9b90a0a51 Mon Sep 17 00:00:00 2001
From: Jarkko Hietaniemi <jhi@iki.fi>
Date: Sat, 26 Apr 2014 14:06:51 -0400
Subject: [PATCH] Fix for Coverity perl5 CIDs 29069, 29070, 29071: Unintended
 sign extension: ... ... if ... U8 (8 bits unsigned) ... 32 bits, signed ...
 64 bits, unsigned ... is greater than 0x7FFFFFFF, the upper bits of the
 result will all be 1.

More casting to make the signed step stay away.
---
 cpan/Encode/Unicode/Unicode.xs |  2 +-
 doop.c                         |  4 ++--
 utf8.c                         | 17 ++++++++++-------
 3 files changed, 13 insertions(+), 10 deletions(-)

diff --git a/cpan/Encode/Unicode/Unicode.xs b/cpan/Encode/Unicode/Unicode.xs
index 4abbf72..78b8513 100644
--- a/cpan/Encode/Unicode/Unicode.xs
+++ b/cpan/Encode/Unicode/Unicode.xs
@@ -80,7 +80,7 @@ enc_unpack(pTHX_ U8 **sp, U8 *e, STRLEN size, U8 endian)
 	if (endian == 'v')
 	    break;
 	v |= (*s++ << 16);
-	v |= (*s++ << 24);
+	v |= (UV)(*s++ << 24);
 	break;
     default:
 	croak("Unknown endian %c",(char) endian);
diff --git a/doop.c b/doop.c
index 5031af8..6814a77 100644
--- a/doop.c
+++ b/doop.c
@@ -846,7 +846,7 @@ Perl_do_vecget(pTHX_ SV *sv, SSize_t offset, int size)
 			((UV) s[uoffset + 1] << 48) +
 			((UV) s[uoffset + 2] << 40) +
 			((UV) s[uoffset + 3] << 32) +
-			(     s[uoffset + 4] << 24);
+			((UV) s[uoffset + 4] << 24);
 		else if (uoffset + 6 >= srclen)
 		    retnum =
 			((UV) s[uoffset    ] << 56) +
@@ -863,7 +863,7 @@ Perl_do_vecget(pTHX_ SV *sv, SSize_t offset, int size)
 			((UV) s[uoffset + 3] << 32) +
 			((UV) s[uoffset + 4] << 24) +
 			((UV) s[uoffset + 5] << 16) +
-			(     s[uoffset + 6] <<  8);
+			((UV) s[uoffset + 6] <<  8);
 	    }
 #endif
 	}
diff --git a/utf8.c b/utf8.c
index fa5b4a7..38641b7 100644
--- a/utf8.c
+++ b/utf8.c
@@ -2809,7 +2809,6 @@ Perl_swash_fetch(pTHX_ SV *swash, const U8 *ptr, bool do_utf8)
     STRLEN slen;
     STRLEN needents;
     const U8 *tmps = NULL;
-    U32 bit;
     SV *swatch;
     const U8 c = *ptr;
 
@@ -2939,17 +2938,21 @@ Perl_swash_fetch(pTHX_ SV *swash, const U8 *ptr, bool do_utf8)
 
     switch ((int)((slen << 3) / needents)) {
     case 1:
-	bit = 1 << (off & 7);
-	off >>= 3;
-	return (tmps[off] & bit) != 0;
+	return ((UV) tmps[off >> 3] & (1 << (off & 7))) != 0;
     case 8:
-	return tmps[off];
+	return ((UV) tmps[off]);
     case 16:
 	off <<= 1;
-	return (tmps[off] << 8) + tmps[off + 1] ;
+	return
+            ((UV) tmps[off    ] << 8) +
+            ((UV) tmps[off + 1]);
     case 32:
 	off <<= 2;
-	return (tmps[off] << 24) + (tmps[off+1] << 16) + (tmps[off+2] << 8) + tmps[off + 3] ;
+	return
+            ((UV) tmps[off    ] << 24) +
+            ((UV) tmps[off + 1] << 16) +
+            ((UV) tmps[off + 2] <<  8) +
+            ((UV) tmps[off + 3]);
     }
     Perl_croak(aTHX_ "panic: swash_fetch got swatch of unexpected bit width, "
 	       "slen=%"UVuf", needents=%"UVuf, (UV)slen, (UV)needents);
-- 
1.9.2


```
</details>

**Comment by p5pRT** [comment]
### From @tonycoz

On Sat Apr 26 13&#8203;:03&#8203;:01 2014\, jhi wrote&#8203;:
> Attached\.

Assuming I understand what Coverity is complaining about\, shouldn't&#8203;:

\-\-\- a/cpan/Encode/Unicode/Unicode\.xs
\+\+\+ b/cpan/Encode/Unicode/Unicode\.xs
@&#8203;@&#8203; \-80\,7 \+80\,7 @&#8203;@&#8203; enc\_unpack\(pTHX\_ U8 \*\*sp\, U8 \*e\, STRLEN size\, U8 endian\)
&nbsp;     if \(endian == 'v'\)
&nbsp;         break;
&nbsp;     v |= \(\*s\+\+ \<\< 16\);
\-    v |= \(\*s\+\+ \<\< 24\);
\+    v |= \(UV\)\(\*s\+\+ \<\< 24\);
&nbsp;     break;
&nbsp;     default&#8203;:
&nbsp;     croak\("Unknown endian %c"\,\(char\) endian\);

be&#8203;:

\+    v |= \(\(UV\)\*s\+\+ \<\< 24\);

Though I think it's incorrect to complain here\, \*s is unsigned char\, with the usual integer promotions that should become unsigned int and the value of the \<\< operator is has the type of its left operand\.

Tony

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @jhi

On Monday\-201404\-28\, 0&#8203;:06\, Tony Cook via RT wrote&#8203;:
> On Sat Apr 26 13&#8203;:03&#8203;:01 2014\, jhi wrote&#8203;:
>> Attached\.
>
> Assuming I understand what Coverity is complaining about\, shouldn't&#8203;:
>
> \-\-\- a/cpan/Encode/Unicode/Unicode\.xs
> \+\+\+ b/cpan/Encode/Unicode/Unicode\.xs
> @&#8203;@&#8203; \-80\,7 \+80\,7 @&#8203;@&#8203; enc\_unpack\(pTHX\_ U8 \*\*sp\, U8 \*e\, STRLEN size\, U8 endian\)
>       if \(endian == 'v'\)
>           break;
>       v |= \(\*s\+\+ \<\< 16\);
> \-    v |= \(\*s\+\+ \<\< 24\);
> \+    v |= \(UV\)\(\*s\+\+ \<\< 24\);
>       break;
>       default&#8203;:
>       croak\("Unknown endian %c"\,\(char\) endian\);
>
> be&#8203;:
>
> \+    v |= \(\(UV\)\*s\+\+ \<\< 24\);

This is the exact error from Coverity&#8203;:

CID 29071&#8203;: Unintended sign extension \(SIGN\_EXTENSION\)
sign\_extension&#8203;: Suspicious implicit sign extension&#8203;: \*s\+\+ with type 
unsigned char \(8 bits\, unsigned\) is promoted in \*s\+\+ \<\< 24 to type int 
\(32 bits\, signed\)\, then sign\-extended to type unsigned long \(64 bits\, 
unsigned\)\. If \*s\+\+ \<\< 24 is greater than 0x7FFFFFFF\, the upper bits of 
the result will all be 1\.

> Though I think it's incorrect to complain here\, \*s is unsigned char\, with the usual integer promotions that should become unsigned int and the value of the \<\< operator is has the type of its left operand\.
>
> Tony
>

**Comment by p5pRT** [comment]
### From @jhi

>
> be&#8203;:
>
> \+    v |= \(\(UV\)\*s\+\+ \<\< 24\);
>
> Though I think it's incorrect to complain here\, \*s is unsigned char\, with the usual integer promotions that should become unsigned int and the value of the \<\< operator is has the type of its left operand\.

Patch refreshed as suggested\, and attached\.

**Comment by p5pRT** [comment]
### From @jhi

<details><summary>0001-Fix-for-Coverity-perl5-CIDs-29069-29070-29071.patch</summary>

```diff
From d39b6115b61a3dc4320c837f65ba97e3323cbe73 Mon Sep 17 00:00:00 2001
From: Jarkko Hietaniemi <jhi@iki.fi>
Date: Sat, 26 Apr 2014 14:06:51 -0400
Subject: [PATCH] Fix for Coverity perl5 CIDs 29069, 29070, 29071: Unintended
 sign extension: ... ... if ... U8 (8 bits unsigned) ... 32 bits, signed ...
 64 bits, unsigned ... is greater than 0x7FFFFFFF, the upper bits of the
 result will all be 1.

More casting to make the signed step stay away.
---
 cpan/Encode/Unicode/Unicode.xs |  2 +-
 doop.c                         |  4 ++--
 utf8.c                         | 17 ++++++++++-------
 3 files changed, 13 insertions(+), 10 deletions(-)

diff --git a/cpan/Encode/Unicode/Unicode.xs b/cpan/Encode/Unicode/Unicode.xs
index 4abbf72..6900fe6 100644
--- a/cpan/Encode/Unicode/Unicode.xs
+++ b/cpan/Encode/Unicode/Unicode.xs
@@ -80,7 +80,7 @@ enc_unpack(pTHX_ U8 **sp, U8 *e, STRLEN size, U8 endian)
 	if (endian == 'v')
 	    break;
 	v |= (*s++ << 16);
-	v |= (*s++ << 24);
+	v |= ((UV)*s++ << 24);
 	break;
     default:
 	croak("Unknown endian %c",(char) endian);
diff --git a/doop.c b/doop.c
index 5031af8..6814a77 100644
--- a/doop.c
+++ b/doop.c
@@ -846,7 +846,7 @@ Perl_do_vecget(pTHX_ SV *sv, SSize_t offset, int size)
 			((UV) s[uoffset + 1] << 48) +
 			((UV) s[uoffset + 2] << 40) +
 			((UV) s[uoffset + 3] << 32) +
-			(     s[uoffset + 4] << 24);
+			((UV) s[uoffset + 4] << 24);
 		else if (uoffset + 6 >= srclen)
 		    retnum =
 			((UV) s[uoffset    ] << 56) +
@@ -863,7 +863,7 @@ Perl_do_vecget(pTHX_ SV *sv, SSize_t offset, int size)
 			((UV) s[uoffset + 3] << 32) +
 			((UV) s[uoffset + 4] << 24) +
 			((UV) s[uoffset + 5] << 16) +
-			(     s[uoffset + 6] <<  8);
+			((UV) s[uoffset + 6] <<  8);
 	    }
 #endif
 	}
diff --git a/utf8.c b/utf8.c
index fa5b4a7..38641b7 100644
--- a/utf8.c
+++ b/utf8.c
@@ -2809,7 +2809,6 @@ Perl_swash_fetch(pTHX_ SV *swash, const U8 *ptr, bool do_utf8)
     STRLEN slen;
     STRLEN needents;
     const U8 *tmps = NULL;
-    U32 bit;
     SV *swatch;
     const U8 c = *ptr;
 
@@ -2939,17 +2938,21 @@ Perl_swash_fetch(pTHX_ SV *swash, const U8 *ptr, bool do_utf8)
 
     switch ((int)((slen << 3) / needents)) {
     case 1:
-	bit = 1 << (off & 7);
-	off >>= 3;
-	return (tmps[off] & bit) != 0;
+	return ((UV) tmps[off >> 3] & (1 << (off & 7))) != 0;
     case 8:
-	return tmps[off];
+	return ((UV) tmps[off]);
     case 16:
 	off <<= 1;
-	return (tmps[off] << 8) + tmps[off + 1] ;
+	return
+            ((UV) tmps[off    ] << 8) +
+            ((UV) tmps[off + 1]);
     case 32:
 	off <<= 2;
-	return (tmps[off] << 24) + (tmps[off+1] << 16) + (tmps[off+2] << 8) + tmps[off + 3] ;
+	return
+            ((UV) tmps[off    ] << 24) +
+            ((UV) tmps[off + 1] << 16) +
+            ((UV) tmps[off + 2] <<  8) +
+            ((UV) tmps[off + 3]);
     }
     Perl_croak(aTHX_ "panic: swash_fetch got swatch of unexpected bit width, "
 	       "slen=%"UVuf", needents=%"UVuf, (UV)slen, (UV)needents);
-- 
1.9.2


```
</details>

**Comment by p5pRT** [comment]
### From @tonycoz

On Mon Apr 28 05&#8203;:14&#8203;:55 2014\, jhi wrote&#8203;:
> >
> > be&#8203;:
> >
> > \+     v |= \(\(UV\)\*s\+\+ \<\< 24\);
> >
> > Though I think it's incorrect to complain here\, \*s is unsigned char\,
> > with the usual integer promotions that should become unsigned int and
> > the value of the \<\< operator is has the type of its left operand\.
> 
> Patch refreshed as suggested\, and attached\.

That looks right\, though the cpan/ changes will need to go through upstream\.

I re\-read the related bits of the spec and I was wrong about the promotions\.

Tony

**Comment by p5pRT** [comment]
### From @jhi

On Tuesday\-201404\-29\, 0&#8203;:54\, Tony Cook via RT wrote&#8203;:
> That looks right\, though the cpan/ changes will need to go through upstream\.

I will let Dan Kogai know\.

**Comment by p5pRT** [comment]
### From @tonycoz

On Mon Apr 28 05&#8203;:14&#8203;:55 2014\, jhi wrote&#8203;:
> >
> > be&#8203;:
> >
> > \+     v |= \(\(UV\)\*s\+\+ \<\< 24\);
> >
> > Though I think it's incorrect to complain here\, \*s is unsigned char\,
> > with the usual integer promotions that should become unsigned int and
> > the value of the \<\< operator is has the type of its left operand\.
> 
> Patch refreshed as suggested\, and attached\.

The cpan/Encode changes were included in c7981a06d5acac597f1bcbdd7664eb04ee6ad5bd \(Encode 2\.60\)\.

The rest of the patch was applied as e7aca353f7af5d2dbed5f0e55c23d87f17c2406c\.

Tony

**Comment by p5pRT** [comment]
@tonycoz - Status changed from 'open' to 'resolved'
